addons/controllers/packageinstallstatus_controller.go

```diff
@@ -0,0 +1,448 @@
+type PackageInstallStatusReconciler struct {
+	Ctx        context.Context
+	Log        logr.Logger
+	Client     client.Client
+	Scheme     *runtime.Scheme
+	controller controller.Controller
+	Tracker    *remote.ClusterCacheTracker
+	// aggregatedAPIResourcesClient is used when it is required to directly read from the server and not to use object caches.
+	aggregatedAPIResourcesClient client.Client
+}
+
+//+kubebuilder:rbac:groups=addons.tanzu.vmware.com,resources=packageinstallstatuses,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=addons.tanzu.vmware.com,resources=packageinstallstatuses/status,verbs=get;update;patch
+
+// Reconcile performs the reconciliation action for the controller; which is reflecting the reconciliation status of each core/additional package
+// into corresponding ClusterBootstrap resource of the cluster
+func (r *PackageInstallStatusReconciler) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	log := r.Log.WithValues(constants.ClusterNamespaceLogKey, req.Namespace, constants.ClusterNameLogKey, req.Name)
+
+	var (
+		clusterClient client.Client
+		clusterRole   ClusterRole
+	)
+
+	// get cluster object
+	cluster := &clusterapiv1beta1.Cluster{}
+	if err := r.Client.Get(ctx, req.NamespacedName, cluster); err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Info("Cluster not found")
+			return ctrl.Result{}, nil
+		}
+		log.Error(err, "unable to fetch cluster")
+		return ctrl.Result{}, err
+	}
+
+	// if cluster is marked for deletion, then no reconciliation is needed
+	if !cluster.GetDeletionTimestamp().IsZero() {
+		return ctrl.Result{}, nil
+	}
+
+	clusterLabels := cluster.GetLabels()
+
+	// make sure the TKR object exists
+	tkrName := clusterLabels[constants.TKRLabelClassyClusters]
+	tkr, err := util.GetTKRByName(ctx, r.Client, tkrName)
+	if err != nil {
+		log.Error(err, "unable to fetch TKR object", "name", tkrName)
+		return ctrl.Result{}, err
+	}
+	// if tkr is not found, should not requeue for the reconciliation
+	if tkr == nil {
+		log.Info("TKR object not found", "name", tkrName)
+		return ctrl.Result{}, nil
+	}
+
+	// determine cluster role
+	_, isManagementCluster := clusterLabels[tkrconstants.ManagementClusterRoleLabel]
+
+	if isManagementCluster {
+		// the cluster is a management cluster
+		clusterRole = clusterRoleManagement
+		clusterClient = r.Client
+	} else {
+		// the cluster is a remote workload cluster
+		clusterRole = clusterRoleWorkload
+		remoteClient, err := r.Tracker.GetClient(ctx, clusterapiutil.ObjectKey(cluster))
+		if err != nil {
+			log.Error(err, "error getting remote cluster's client")
+			return ctrl.Result{}, err
+		}
+		clusterClient = remoteClient
+		log.Info("successfully got remoteClient")
+		// set watch if not already set. If the watch already exists, it doesn't get re-created
+		if err := r.watchPackageInstalls(ctx, cluster, log); err != nil {
```

Comment:
Question: What about clusters that are deleted? Do we drop the created trackers for those somewhere?

Reply:
```deleteAccessor``` in CAPI is un-exported, but is called inside the ```healthCheckCluster``` goroutine  which is automatically called once I start the remote watch, when the cluster is eventually not found.

Reply:
@sbueringer has mentioned in the [following comment ](https://github.com/vmware-tanzu/tanzu-framework/pull/1781#discussion_r832198751)that can use ```ClusterCacheReconciler``` too which does some best effort deletion, but might not be worth it to add it.

Reply:
I tested the ClusterCacheReconciler a bit and I would recommend using it now :).

Essentially the behavior is:
* After a Cluster is deleted the CluterCacheReconciler drops the accessor via deleteAccessor 
    * deleted in the sense of that the resource is actually deleted from the apiserver/etcd not only that the deletionTimestamp is set
* As fallback after deletion or if the actual cluster goes down for some reason the healthchecking inside of the ClusterCacheTracker also deletes the accessor

Reply:
Thanks for this. Adding ```CluterCacheReconciler``` now